### ISSUE TITLE
skip/include directives do not `avoidOptionals` if field type is a list

### ISSUE THREAD

**Issue by brainkim** [dotansimha/graphql-code-generator#6748]
**Describe the bug**

The `@skip` and `@include` directives seem to be generating optional types for `GraphQLList` fields, even when `avoidOptionals` is true.

**To Reproduce**
https://codesandbox.io/s/practical-swanson-b93wg

1. My GraphQL schema:

```graphql
type Query {
  me: User!
}

type User {
  messages: [Message!]!
}

type Message {
  content: String!
}
```

2. My GraphQL operations:

```graphql
query MyQuery($include: Boolean!) {
  me {
    messages @include(if: $include) {
      content
    }
  }
}
```

3. My `codegen.yml` config file:

```yml
schema: schema.graphql
documents: document.graphql
generates:
  types.ts:
    plugins:
      - typescript
      - typescript-operations
config:
  avoidOptionals: true
  nonOptionalTypename: true
```

**Expected behavior**

I expect TypeScript optionals not to be included in the generated output.

**Actual behavior**
```ts
export type MyQueryQuery = (
  { __typename: 'Query' }
  & { me: (
    { __typename: 'User' }
    & { messages?: Maybe<(
      { __typename: 'Message' }
      & Pick<Message, 'content'>
    )> }
  ) }
);
```

`messages` is optional.

**Environment:**

- OS: MacOS Big Sur (11.6)
- "@graphql-codegen/add": "2.0.2",
- "@graphql-codegen/cli": "1.20.1",
- "@graphql-codegen/typescript": "1.21.0",
- "@graphql-codegen/typescript-operations": "1.17.14",
- "graphql": "15.5.0"

**Additional context**

Probably related to this PR: 
https://github.com/dotansimha/graphql-code-generator/pull/5104

**Comment by gilgardosh** [comment]
Hi @brainkim ,
could you test it with this alphas, and confirm it fixes your issue?
https://github.com/dotansimha/graphql-code-generator/pull/6752#issuecomment-930122239

**Comment by brainkim** [comment]
I tried upgrading typescript-operations to 2.1.5, but am now running into new errors related to the `preResolveTypes` option erasing `Maybe` types in an unexpected way. Looking into it.

Hello from ApolloGraphQL💜

**Comment by gilgardosh** [comment]
Rethinking this case.
If conditional directive is used, the expected outcome must be optional, even if `avoidOptinals` is set.
case include is `false`, `messages` property shouldn't be in the response.
so:

`export type MyQueryQuery = (
  { __typename: 'Query' }
  & { me: (
    { __typename: 'User' }
    & { messages?: Array<(
      { __typename: 'Message' }
      & Pick<Message, 'content'>
    )> }
  ) }
);` 

@brainkim  what do you think?

**Comment by dotansimha** [comment]
I agree with @gilgardosh and @n1ru4l in this case. 
`skip` and `include` will cause the field to either be there or not (the actual key in the response), so TS optional seems the correct way to represent it. `avoidOptionals` it a matter of choice, but in this case, it might create an inaccurate TS type...

**Comment by n1ru4l** [comment]
Just to add to that another way of representing this (without optionals could be the following):

```ts
export type MyQueryQuery = { __typename: "Query" } & {
  me: { __typename: "User" } & (
    | {}
    | { messages: Array<{ __typename: "Message" } & { content: string }> });
};

const foo = (d: MyQueryQuery) => {
  // @ts-expect-error
  d.me.messages;
  if ("messages" in d.me) {
    // do sth with it here
    d.me.messages;
  }
};
```

[on TS Playground](https://www.typescriptlang.org/play?ssl=25&ssc=2&pln=1&pc=1#code/KYDwDg9gTgLgBDAnmYcCyiCKBXYUu75wC8cAFAFBxwDecA+vUigHYCGAtsAFxwDkOPIj5wAvlTgAyWnC69K1anUbNg7OfwCqAZzwjxiqeQmGaouCcUAfGV23a2Ac2DbeAQShQ2iADwLDAcpMyGqcPPxoLg7O+pYBUnQAxhAsMGowvNowUACWLI5icYoAlAB8RQYlEsWFxQDcFI0UySxZcABmEBAk5AAmvBiC+EOINcSltBIA9FNwAAIw2gC0oCiJMCue0BK9AHRc+1FOLhI57eQARHbRLhdweXB7XDU0RTOP3VkAFnAA7jkwH4AuBfPDAIpPYCHezHbQScSiIA)

This would be beneficial in a case where the boolean affects multiple fields🤔

```graphql
query MyQuery($include: Boolean!) {
  message {
    id @include(if: $include)
    content @include(if: $include)
  }
}
```

```ts
export type MessageQuery = { __typename: "Query" } & {
  message:
    | ({ __typename: "Message" } & ({} | { id: string }) &
        ({} | { content: string }))
    | null;
};

const foo1 = (d: MessageQuery) => {
  if (!d.message) return;

  // @ts-expect-error
  d.message.id;
  // @ts-expect-error
  d.message.content;

  if ("id" in d.message) {
    // do sth with it here
    d.message.id;
  }
  if ("content" in d.message) {
    // do sth with it here
    d.message.content;
  }
};
```

Or with an inline fragment:

```graphql
query MyQuery($include: Boolean!) {
  message {
    ... on Message(if: $include) {
    id
    content
  }
}

```

```ts
export type MessageQuery = { __typename: "Query" } & {
  message:
    | ({ __typename: "Message" } & ({} | { id: string; content: string }))
    | null;
};

const foo1 = (d: MessageQuery) => {
  if (!d.message) return;

  // @ts-expect-error
  d.message.id;
  // @ts-expect-error
  d.message.content;

  if ("id" in d.message) {
    // do sth with it here
    d.message.id;
    d.message.content;
  }
};
```

**Comment by brainkim** [comment]
Ah yeah. Some disorganized thoughts.

1. The `avoidOptionals` option is frustrating in the sense that for input types I want optional nullable fields, and for output types I don’t. I share your intuitions that there is a “correct” generated output here, and that `@skip`/`@include` directives should be modeled with TypeScript optionals, but I think it’s confusing to have optional properties continue to appear in generated code even if you set `avoidOptionals` to `true`. I know and appreciate that graphql-code-generator tries to be supremely configurable, but I think `avoidOptionals` is one of those things which I wish I didn’t have control over.
2. TypeScript is undergoing a campaign to distinguish “missing” or “optional” properties from properties which can be explicitly assigned `undefined` (https://github.com/microsoft/TypeScript/pull/43947 which implements the `--strictOptionalProperties` option). I’m not sure how this affects codegen decisions surrounding optionals yet, but I’m sure it does somehow.
3. I dislike representing skipped/included fragments as `T & ({} | TFragment)` as @n1ru4l suggests because it makes it difficult to use indexed access types to access nested object or interface fields ([typescript playground example](https://www.typescriptlang.org/play?ssl=8&ssc=61&pln=1&pc=1#code/KYDwDg9gTgLgBDAnmYcCyiCKBXYUu75wC8cA3nAPqVIoB2AhgLbABccARDnoh3AL5wAZOQBQcOC3YVqtYIymcAqgGc8fQSIAU4iXAA+5fromGKLFSoYBzYCvYBBKFAaIAPDJrJ5zNpzR2VrYawuRwAMYQdDDyMOwqMFAAlnTWAgB8AgCUANyi-HmionLoBDzc+JQWQXYkpRVl+ADaAOQsLQC6rdU2dp05QA)). This is kind of important for Apollo because without the ability to reference subfields, migrating from apollo codegen to graphql code generator is going to be difficult. I also dislike using the `in` operator to duck-type, but that’s another matter.
4. Any decision we make about skipped/included fragments should likely be mirrored for modeling `@defer` in operations.


Thanks for the responses!

**Comment by n1ru4l** [comment]
Regarding 1: We are already considering dropping the `avoidOptionals` option in a future major release. For now the avoidOptionals should be adjusted to avoid optionals where it makes sense, like @gilgardosh mentioned

Regarding 3: This is something we could make configurable.

Regarding 4: skip/include and defer on fragments are IMHO best modeled with the `T & ({} | TFragment)` approach as they allow checking for a single property with `property in obj` in order to discriminate the type correctly, instead of checking for every single one.

**Comment by dotansimha** [comment]
Just joining this super interesting discussion. 

I think the path of TS optional is clear - it makes things more complex and could be solved with a simple explicit `undefined` type. 
But, as @n1ru4l said, optionals are important and needed in cases where fields might really not exists at all. Mainly because of `in` operator that could give false positives with `undefined` values. 

```
> const a = { };
> 'b' in a
false
> const a2 = { b: undefined };
> 'b' in a2
true
```

With `skip` and `include`, GraphQL execution layer will not return a field at all when a field was omitted due to a truthy evaluation of the directive, which means that the field is really is optional, from the point of view of the consumer. 

When we started working on it, we tried to align the `avoidOptionals` behaviour and make it consistent as we do in other places, but now it seems like optional in this case might be important in order to reflect the actual possible responses from the GraphQL response.

@brainkim what do you think?

**Comment by brainkim** [comment]
My colleague @martinbonnin mentioned a fun little example:

> ```graphql
> type Query {
>   me: User!
> }
> 
> type User {
>   message: Message
> }
> 
> type Message {
>   content: String!
> }
> query MyQuery($include: Boolean!) {
>   me {
>     message @include(if: $include) {
>       content
>     }
>   }
> }
> ```
> 
> Should message be optional and nullable or just optional or just nullable ?

Another wrinkle to think about.

**Comment by n1ru4l** [comment]
@brainkim Forthat example, the generated code should be:

```ts
export type MyQueryQuery = { __typename: "Query" } & {
  me: { __typename: "User" } & {
    messages?: Array<
      { __typename: "Message" } & Pick<Message, "content">
    > | null;
  };
};
```

**Comment by brainkim** [comment]
@n1ru4l I took a brief look at the PR referenced in the CHANGELOG.md. [This test](https://github.com/dotansimha/graphql-code-generator/commit/8a576b49ae0a772de637377c489d77839ab54656#r58011622) seems to indicate that we’re respecting `avoidOptionals` and making `user` and `group` non-optional? I’m probably just misreading the code 😅

For posterity, can you provide a brief description of what y‘all decided on with regard to `avoidOptionals` and inclusion directives? If there are unexpected types I can reopen this issue or create a new one.

**Comment by gilgardosh** [comment]
@brainkim Thanks for being thorough! there actually is an issue with inline fragments mixed with conditional directive such as in this case.

The directive is not influencing `user` nor `group` directly, but the query type as a hole.
For example, result cannot be 
```
{
  user: "userName"
}
```
since `group` is required if result exists.

The `Maybe` wrapping `user` is due to it's own definition.

Here is a [fix](https://github.com/dotansimha/graphql-code-generator/pull/6852), which you mentioned is not your cup of tea, but is a true representation of this case.

TL;DR:
`<CurrentType> | {}`
